### PR TITLE
TESTCASES: ec_tests: Support ECDH KATs for secure key tokens

### PR DESCRIPTION
**Pull request by ifranzki**
For secure key tokens, the derived secret can not be compared with the expected key value for known answer tests (KATs), because the derived secret is a secure key that never reveals its key value in clear.

However, the test can calculate an HMAC of some know data using the derived key as well as with the expected key value from the KAT. If both keys are the same, then the calculated HMAC must also be the same.

### CODE REVIEW

File: testcases/crypto/ec_func.c [hunk 1]
```diff
@@ -362,6 +364,50 @@ CK_RV run_HMACSign(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE h_key,
...
+    default:
+        testcase_notice("Mechanism %s is not supported. Skipping key check",
+                        p11_get_ckm(&mechtable_funcs, hmac_mech));
+        *mac_len = 0;
+        return CKR_OK;
+    }
+
+    out_len = *mac_len;
+    if (HMAC(md, key, key_len, data, sizeof(data), mac, &out_len) == NULL) {
+        testcase_notice("HMAC (OpneSSL) failed");
```
Comment (jschmidb) [inline testcases/crypto/ec_func.c:403]: OpenSSL

File: testcases/crypto/ec_func.c [hunk 2]
```diff
@@ -1207,51 +1296,104 @@ CK_RV run_DeriveECDHKeyKAT(void)
...
+
+            rc = run_HMACSign(session, secret_keyA, ecdh_tv[i].derived_key_len,
+                              CKM_SHA_1_HMAC, mac1, &mac1_len);
+            if (rc != CKR_OK) {
+                testcase_fail("HMAC for derived key #1 failed: %s",
+                              p11_get_ckr(rc));
+                goto testcase_cleanup;
+            }
+
+            if (mac1_len != 0 && /* skip chck if mac can't be calcualted */
```
Comment (jschmidb) [inline testcases/crypto/ec_func.c:1374]: skip check if mac can't be calculated

Comment (jschmidb) [reply testcases/crypto/ec_func.c:1374]: ... also below

File: testcases/crypto/ec_func.c (hunk 1 above)
Comment (ifranzki) [reply testcases/crypto/ec_func.c:403]: fixed

File: testcases/crypto/ec_func.c (hunk 2 above)
Comment (ifranzki) [reply testcases/crypto/ec_func.c:1373]: fixed both

**Review by jschmidb — APPROVED** [review]
lgtm

**Comment by ifranzki** [comment]
Added another commit to support compressed or hybrid EC points with EP11 as public data with ECDH. This also removes the ugly checking in the testcases for such EC points.
@jschmidb please take one more look.

File: usr/lib/ep11_stdll/ep11_specific.c [hunk 3]
```diff
@@ -6282,6 +6282,8 @@ CK_RV ep11tok_derive_key(STDLL_TokData_t *tokdata, SESSION *session,
          * The EP11 host library only accepts the raw form, thus convert
          * it to the raw format if the caller specified it in the DER-encoded
          * form.
+         * Also, the EP11 host library only accepts EC points is uncompressed
```
Comment (jschmidb) [inline usr/lib/ep11_stdll/ep11_specific.c:6285]: in compressed form

File: usr/lib/common/mech_ec.c [hunk 4]
```diff
@@ -1530,3 +1530,73 @@ int ec_point_from_public_data(const CK_BYTE *data, CK_ULONG data_len,
...
+        break;
+    }
+
+    if (rc != CKR_OK && *allocated) {
+        free(*ec_point);
+        *ec_point = NULL;
+        *ec_point_len = 0;
+        *allocated = FALSE;
+    }
+    if (ec_point2 != NULL)
```
Comment (jschmidb) [inline usr/lib/common/mech_ec.c:1598]: Hmm, in line 1581 *ec_point = ec_point2;
If something goes wrong, i.e. rc != CKR_OK, *ec_point gets freed, which is ok.
But also in the good case, in line 1599, ec_point2 is freed, which means *ec_point is freed. Is this intended?

File: usr/lib/ep11_stdll/ep11_specific.c (hunk 3 above)
Comment (ifranzki) [reply usr/lib/ep11_stdll/ep11_specific.c:6285]: fixed

File: usr/lib/common/mech_ec.c (hunk 4 above)
Comment (ifranzki) [reply usr/lib/common/mech_ec.c:1598]: > But also in the good case, in line 1599, ec_point2 is freed

No it is not. ec_point2 is set to NULL in line 1584, so it won't be freed.

**Review by jschmidb — APPROVED** [review]

**Pull request merged**